ksqldb-execution/src/main/java/io/confluent/ksql/execution/interpreter/ExpressionInterpreter.java

```diff
@@ -0,0 +1,769 @@
+    public Pair<Object, SqlType> visitNullLiteral(final NullLiteral node, final Void context) {
+      return new Pair<>(null, null);
+    }
+
+    @Override
+    public Pair<Object, SqlType> visitLambdaExpression(
+        final LambdaFunctionCall lambdaFunctionCall, final Void context) {
+      return visitUnsupported(lambdaFunctionCall);
+    }
+
+    @Override
+    public Pair<Object, SqlType> visitLambdaVariable(
+        final LambdaVariable lambdaLiteral, final Void context) {
+      return visitUnsupported(lambdaLiteral);
+    }
+
+    @Override
+    public Pair<Object, SqlType> visitUnqualifiedColumnReference(
+        final UnqualifiedColumnReferenceExp node,
+        final Void context
+    ) {
+      final Column schemaColumn = schema.findValueColumn(node.getColumnName())
+          .orElseThrow(() ->
+              new KsqlException("Field not found: " + node.getColumnName()));
+
+      final Object value = row.get(schemaColumn.index());
+
+      return new Pair<>(value, schemaColumn.type());
+    }
+
+    @Override
+    public Pair<Object, SqlType> visitQualifiedColumnReference(
+        final QualifiedColumnReferenceExp node,
+        final Void context
+    ) {
+      throw new UnsupportedOperationException(
+          "Qualified column reference must be resolved to unqualified reference before codegen"
+      );
+    }
+
+    @Override
+    public Pair<Object, SqlType> visitDereferenceExpression(
+        final DereferenceExpression node, final Void context
+    ) {
+      final SqlType functionReturnSchema = expressionTypeManager.getExpressionSqlType(node);
+
+      final Pair<Object, SqlType> base = process(node.getBase(), context);
+
+      if (base.getRight().baseType() != SqlBaseType.STRUCT) {
+        throw new KsqlException("Can only dereference Struct type, instead got " + base.getRight());
+      }
+
+      final Struct struct = (Struct) base.getLeft();
+      final String field = (String) process(new StringLiteral(node.getFieldName()), context)
+          .getLeft();
+      final Object dereference = struct.get(field);
+
+      return new Pair<>(dereference, functionReturnSchema);
+    }
+
+    public Pair<Object, SqlType> visitLongLiteral(final LongLiteral node, final Void context) {
+      return new Pair<>(node.getValue(), SqlTypes.BIGINT);
+    }
+
+    @Override
+    public Pair<Object, SqlType> visitIntegerLiteral(
+        final IntegerLiteral node,
+        final Void context
+    ) {
+      return new Pair<>(node.getValue(), SqlTypes.INTEGER);
+    }
+
+    @Override
+    public Pair<Object, SqlType> visitFunctionCall(final FunctionCall node, final Void context) {
+      final UdfFactory udfFactory = functionRegistry.getUdfFactory(node.getName());
+      final List<SqlType> argumentSchemas = node.getArguments().stream()
+          .map(expressionTypeManager::getExpressionSqlType)
+          .collect(Collectors.toList());
+
+      final KsqlScalarFunction function = udfFactory.getFunction(argumentSchemas);
```

Comment:
Should we catch/throw inside this visiting call if we cannot find a matching function signature, or is it handled outside of this function call already?